servant-server/src/Servant/Server/Internal.hs

```diff
@@ -405,6 +406,48 @@ instance (KnownSymbol sym, HasServer api context)
           examine v | v == "true" || v == "1" || v == "" = True
                     | otherwise = False
 
+-- | If you use @'QueryParamForm' BookSearchParams@ in one of the endpoints for your API,
+-- this automatically requires your server-side handler to be a function
+-- that takes an argument of type @['BookSearchParams']@.
+--
+-- This lets servant worry about all key-values in the query string
+-- and turning each of them into a value of the type you specify.
+--
+-- You can control how the individual values are converted from 'BookSearchParams'
+-- to your type by simply providing an instance of 'FromForm' for your type.
+--
+-- Example:
+--
+-- > data BookSearchParams = BookSearchParams
+-- >   { title :: Text
+-- >   { authors :: [Text]
+-- >   , page :: Maybe Int
+-- >   } deriving (Generic)
+-- > instance FromForm BookSearchParams
+-- >
+-- > type MyApi = "books" :> QueryParamForm BookSearchParams :> Get '[JSON] [Book]
+-- >
+-- > server :: Server MyApi
+-- > server = getBooksBy
+-- >   where getBooksBy :: BookSearchParams -> Handler [Book]
+-- >         getBooksBy searchParams = ...return all books by these conditions...
+
+instance (FromForm a, HasServer api context)
+      => HasServer (QueryParamForm a :> api) context where
+
+  type ServerT (QueryParamForm a :> api) m =
+    a -> ServerT api m
+
+  route Proxy context subserver = route (Proxy :: Proxy api) context $
+      subserver `addParameterCheck` withRequest paramsCheck
+    where
+      paramsCheck req =
+          case urlDecodeAsForm (BL.drop 1 . BL.fromStrict $ rawQueryString req) of
```

Comment:
I guess `BL.drop 1` is about dropping leading `?`?
Can `rawQueryString` contain extra info (e.g. HTML anchors?), would it ruin parsing?